### DISCUSSION TITLE
Adapt `matplotlib-pgf` example to image given by imshow ?

### DISCUSSION THREAD

**Discussion by jowe19** [tndrle/PyLuaTeX#20] [Ideas]
Hi, 
It's all in the title, I think. Currently it is possible to add graphs without external files using the `io` module. Is it possible to do the same with images from `imshow`? 

I tried some naive things but it doesn't work (and I don't know io at all). 

Thanks for your help. 

```
\documentclass{article}

\usepackage{pyluatex}
\usepackage{unicode-math}
\usepackage{pgf}
\usepackage{url}

\title{PyLuaTeX Example -- Matplotlib PGF}
\author{Tobias Enderle}

\begin{document}

\maketitle

In this document we demonstrate how \emph{matplotlib} plots can be generated and
included in a document. In the Python code in this document, the plot is
configured and saved to PGF\footnote{\url{https://ctan.org/pkg/pgf}} code.
The PGF code is then included in the document and creates the plot.

To avoid intermediate files, the PGF code is saved into an in-memory file
using the Python \verb|io.StringIO| class%
\footnote{\url{https://docs.python.org/3/library/io.html#io.StringIO}}.

\begin{center}
\begin{python}
# example based on
# https://matplotlib.org/stable/gallery/lines_bars_and_markers/simple_plot.html
import matplotlib.pyplot as plt
import numpy as np
import io

Im = plt.imread('crayons.jpg') # un tableau numpy
plt.imshow(Im[:,:,1], cmap = 'gray')

with io.StringIO() as file:  # save to "in-memory file"
    plt.savefig(file, format='pgf', bbox_inches='tight', pad_inches=0.1)
    print(file.getvalue())
\end{python}
\end{center}

For an alternative approach of creating plots with \emph{matplotlib},
consider the example \verb|matplotlib-external.tex|.

\end{document}
```
![crayons](https://user-images.githubusercontent.com/19306526/221028372-0922677d-08fd-4e7b-a5bd-ecf9fd1ee11e.jpg)

**Comment by tndrle** [comment]
Hi,

this is possible with PGF because PGF is basically text commands LaTeX understands. So you can include them in the document directly. Are you able to output PGF with `imshow`? If not, what image format does it generate?

If the zip file you included contains anything crucial, please try to include it in your post directly. I will not open a zip file.

Best

**Reply by jowe19** [reply]
Hi, 
Thank you for your reply. I updated my message, with an example of a tex file and an image, out of the zip.

**Reply by tndrle** [reply]
Hi, thanks for extending your post. When I compile it, I get the Python error
> ValueError: streamed pgf-code does not support raster graphics, consider using 
the pgf-to-pdf option

I'm not an expert on this, but I guess PGF output is not supported for raster graphics.
PGF is for vector graphics. I don't see a way of including raster graphics into LaTeX
without external files.

See also
https://tex.stackexchange.com/questions/47341/can-you-embed-an-image-so-it-doesnt-need-to-be-stored-as-a-separate-file
and
https://tex.stackexchange.com/questions/162013/can-i-bundle-images-in-the-source-file?noredirect=1&lq=1
